core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+
+  try_result try_read_lock()
+  {
+    auto locked = shared.try_lock();
+    if (!locked) { return failed_to_acquire_lock{}; }
+    if (locked.value()->out_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to read from channel that was closed from the read side.\n");
+      return channel_closed{};
+    }
+    if (locked.value()->data.size() == 0) {
+      if (locked.value()->in_dead) {
+        return channel_closed{};
+      } else {
+        return failed_to_acquire_lock{};
+      }
+    }
+
+    return try_result(std::in_place_type<handle>, std::move(locked).value(),
+                      &out_update);
+  }
+
+  try_result try_write_lock()
+  {
+    auto locked = shared.try_lock();
+    if (!locked) { return failed_to_acquire_lock{}; }
+    if (locked.value()->in_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to write to channel that was closed from the write side.\n");
+      return channel_closed{};
+    }
+    if (locked.value()->out_dead) { return channel_closed{}; }
+    if (locked.value()->data.size() >= max_size) {
+      return failed_to_acquire_lock{};
+    }
+
+    return try_result(std::in_place_type<handle>, std::move(locked).value(),
+                      &in_update);
+  }
+
+  void close_in()
+  {
+    shared.lock()->in_dead = true;
+    in_update.notify_one();
+  }
+
+  void close_out()
+  {
+    shared.lock()->out_dead = true;
+    out_update.notify_one();
+  }
+};
+
+template <typename T> class in {
+  std::shared_ptr<queue<T>> shared;
+  bool did_close{false};
+
+ public:
+  in(std::shared_ptr<queue<T>> shared) : shared{std::move(shared)} {}
+  in(in&&) = default;
+  in& operator=(in&&) = default;
+  in(const in&) = delete;
+  in& operator=(const in&) = delete;
+
+  template <typename... Args> bool emplace(Args... args)
+  {
+    if (did_close) { return false; }
+    if (auto handle = shared->write_lock()) {
+      handle->data().emplace_back(std::forward<Args>(args)...);
+      return true;
+    } else {
+      close();
+      return false;
+    }
+  }
+
+  template <typename... Args> bool try_emplace(Args... args)
+  {
+    if (did_close) { return false; }
+    auto result = shared->try_write_lock();
+    if (std::holds_alternative<failed_to_acquire_lock>(result)) {
+      return false;
+    } else if (std::holds_alternative<channel_closed>(result)) {
+      close();
+      return false;
+    } else {
+      std::get<typename queue<T>::handle>(result).data().emplace_back(
+          std::forward<Args>(args)...);
+      return true;
+    }
+  }
+
+  void close()
+  {
+    if (!did_close) {
+      shared->close_in();
+      did_close = true;
+    }
+  }
+
+  bool closed() const { return did_close; }
+
+  ~in()
+  {
+    if (shared) { close(); }
+  }
+};
+
+template <typename T> class out {
```

Comment:
same as above, maybe this should be `output` instead of `out`.

Reply:
done